File: typescript/src/giftcard/api/giftCardApi.ts
// API pubbliche del GiftCard Bounded Context.
// Violazione didattica: l'API dipende direttamente dall'infrastructure repository.

import express, { type Express, type Request, type Response } from 'express';
import { getErrorMessage } from '@/common/api/errorMessage.js';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { GiftCardId } from '../domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '../infrastructure/sqliteGiftCardRepository.js';
import { GiftCardIssuing } from '../application/usecases/giftCardIssuing.js';
import { TopUpRequesting } from '../application/usecases/topUpRequesting.js';
import { requestGiftCardTopUp } from '../application/commands/requestGiftCardTopUp.js';
import { issueGiftCard } from '../application/commands/issueGiftCard.js';
import { parseRequestTopUpRequest } from './requestTopUpRequest.js';
import { toGiftCardResponse } from './giftCardResponse.js';

export class GiftCardApi {
  private readonly _giftCardIssuing: GiftCardIssuing;
  private readonly _giftCardRepository: SqliteGiftCardRepository;
  private readonly _topUpRequesting: TopUpRequesting;

  constructor(giftCardIssuing: GiftCardIssuing, giftCardRepository: SqliteGiftCardRepository, topUpRequesting: TopUpRequesting) {
    this._giftCardIssuing = giftCardIssuing;
    this._giftCardRepository = giftCardRepository;
    this._topUpRequesting = topUpRequesting;
  }

  configure(app: Express): void {
    const router = express.Router();
    router.use(express.json());

    router.post('/gift-cards', (req, res) => this.issue(req, res));
    router.post('/gift-cards/:id/top-up', (req, res) => this.requestTopUp(req, res));
    router.get('/gift-cards/:id', (req, res) => this.getById(req, res));

    app.use(router);
  }

  /**
   * @swagger
   * /gift-cards:
   *   post:
   *     summary: Emite una nuova gift card
   *     tags: [GiftCard]
   *     responses:
   *       201:
   *         description: Gift card emessa
   *         headers:
   *           Location:
   *             description: URI della gift card creata
   *             schema:
   *               type: string
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/GiftCardResponse'
   *       400:
   *         description: Richiesta non valida
   */
  private issue(_req: Request, res: Response): void {
    const cardId = generateId((value) => new GiftCardId(value));

    try {
      this._giftCardIssuing.invoke(issueGiftCard(cardId));
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    const card = this._giftCardRepository.findById(cardId);
    if (card === null) {
      res.status(500).send('gift card not found after issuing');
      return;
    }

    res.status(201)
      .set('Location', `/gift-cards/${cardId.value.value}`)
      .json(toGiftCardResponse(card));
  }

  /**
   * @swagger
   * /gift-cards/{id}/top-up:
   *   post:
   *     summary: Richiede la ricarica di una gift card
   *     tags: [GiftCard]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: string
   *           format: uuid
   *     requestBody:
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/RequestTopUpRequest'
   *     responses:
   *       200:
   *         description: Ricarica richiesta
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/GiftCardResponse'
   *       400:
   *         description: Richiesta non valida
   *       404:
   *         description: Gift card non trovata
   */
  private requestTopUp(req: Request, res: Response): void {
    const cardId = this.parseGiftCardId(req, res);
    if (cardId === null) {
      return;
    }

    let request;
    try {
      request = parseRequestTopUpRequest(req.body);
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    let amount: Money;
    try {
      amount = new Money(Number(request.amount));
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'invalid amount'));
      return;
    }

    try {
      this._topUpRequesting.invoke(requestGiftCardTopUp(cardId, amount));
    } catch (error) {
      if (error instanceof Error && error.message.includes('Gift card not found')) {
        res.status(404).send();
        return;
      }
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    const card = this._giftCardRepository.findById(cardId);
    if (card === null) {
      res.status(500).send('gift card not found after top up');
      return;
    }

    res.status(200).json(toGiftCardResponse(card));
  }

  /**
   * @swagger
   * /gift-cards/{id}:
   *   get:
   *     summary: Recupera una gift card per ID
   *     tags: [GiftCard]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: string
   *           format: uuid
   *     responses:
   *       200:
   *         description: Gift card trovata
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/GiftCardResponse'
   *       400:
   *         description: ID non valido
   *       404:
   *         description: Gift card non trovata
   */
  private getById(req: Request, res: Response): void {
    const id = this.parseGiftCardId(req, res);
    if (id === null) {
      return;
    }

    const card = this._giftCardRepository.findById(id);
    if (card === null) {
      res.status(404).send();
      return;
    }

    res.status(200).json(toGiftCardResponse(card));
  }

  private parseGiftCardId(req: Request, res: Response): GiftCardId | null {
    const idParam = req.params.id;
    if (typeof idParam !== 'string') {
      res.status(400).send('Invalid gift card id format');
      return null;
    }
    try {
      return new GiftCardId(Uuid.fromString(idParam));
    } catch (_error) {
      res.status(400).send('Invalid gift card id format');
      return null;
    }
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/api/GiftCardApi.java
package it.giannibombelli.wsc2026.giftcard.api;

import io.javalin.config.JavalinConfig;
import io.javalin.http.Context;
import io.javalin.openapi.*;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.module.WebApi;
import it.giannibombelli.wsc2026.giftcard.application.commands.IssueGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.commands.RequestGiftCardTopUp;
import it.giannibombelli.wsc2026.giftcard.application.query.GiftCardDetails;
import it.giannibombelli.wsc2026.giftcard.application.query.GiftCardQueryService;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardIssuing;
import it.giannibombelli.wsc2026.giftcard.application.usecases.TopUpRequesting;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

import java.util.UUID;

import static io.javalin.apibuilder.ApiBuilder.get;
import static io.javalin.apibuilder.ApiBuilder.post;

public final class GiftCardApi implements WebApi {
    private final GiftCardIssuing giftCardIssuing;
    private final GiftCardQueryService giftCardQueryService;
    private final TopUpRequesting topUpRequesting;

    public GiftCardApi(GiftCardIssuing giftCardIssuing, GiftCardQueryService giftCardQueryService, TopUpRequesting topUpRequesting) {
        this.giftCardIssuing = giftCardIssuing;
        this.giftCardQueryService = giftCardQueryService;
        this.topUpRequesting = topUpRequesting;
    }

    @Override
    public void configure(JavalinConfig config) {
        config.routes.apiBuilder(() -> {
            post("/gift-cards", this::issue);
            post("/gift-cards/{id}/top-up", this::requestTopUp);
            get("/gift-cards/{id}", this::getById);
        });
    }

    @OpenApi(
        path = "/gift-cards",
        methods = HttpMethod.POST,
        summary = "Issue a new gift card",
        tags = {"GiftCard"},
        responses = {
            @OpenApiResponse(status = "201", content = {@OpenApiContent(from = GiftCardResponse.class)}),
            @OpenApiResponse(status = "400")
        }
    )
    private void issue(Context ctx) {
        GiftCardId cardId = EntityId.generate(GiftCardId::new);

        try {
            IssueGiftCard cmd = new IssueGiftCard(cardId);
            giftCardIssuing.invoke(cmd);
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        GiftCardResponse response = loadGiftCardResponse(ctx, cardId);
        if (response == null) {
            return;
        }

        ctx.status(201)
            .header("Location", "/gift-cards/" + cardId.value())
            .json(response);
    }

    @OpenApi(
        path = "/gift-cards/{id}/top-up",
        methods = HttpMethod.POST,
        summary = "Request a top-up for a gift card",
        tags = {"GiftCard"},
        pathParams = {@OpenApiParam(name = "id", required = true)},
        requestBody = @OpenApiRequestBody(content = {@OpenApiContent(from = RequestTopUpRequest.class)}),
        responses = {
            @OpenApiResponse(status = "200", content = {@OpenApiContent(from = GiftCardResponse.class)}),
            @OpenApiResponse(status = "400"),
            @OpenApiResponse(status = "404")
        }
    )
    private void requestTopUp(Context ctx) {
        GiftCardId cardId = parseGiftCardId(ctx);
        if (cardId == null) {
            return;
        }

        RequestTopUpRequest request = parseRequestTopUpRequest(ctx);
        if (request == null) {
            return;
        }

        if (request.amount() == null) {
            ctx.status(400).result("amount is required");
            return;
        }

        Money amount;
        try {
            amount = new Money(request.amount());
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        try {
            RequestGiftCardTopUp command = new RequestGiftCardTopUp(cardId, amount);
            topUpRequesting.invoke(command);
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        } catch (IllegalStateException ex) {
            ctx.status(404);
            return;
        }

        GiftCardResponse response = loadGiftCardResponse(ctx, cardId);
        if (response == null) {
            return;
        }

        ctx.status(200).json(response);
    }

    @OpenApi(
        path = "/gift-cards/{id}",
        methods = HttpMethod.GET,
        summary = "Get a gift card by id",
        tags = {"GiftCard"},
        pathParams = {@OpenApiParam(name = "id", required = true)},
        responses = {
            @OpenApiResponse(status = "200", content = {@OpenApiContent(from = GiftCardResponse.class)}),
            @OpenApiResponse(status = "400"),
            @OpenApiResponse(status = "404")
        }
    )
    private void getById(Context ctx) {
        GiftCardId id = parseGiftCardId(ctx);
        if (id == null) {
            return;
        }

        var cardOpt = giftCardQueryService.findById(id);
        if (cardOpt.isEmpty()) {
            ctx.status(404);
            return;
        }

        ctx.status(200).json(toResponse(cardOpt.get()));
    }

    private GiftCardId parseGiftCardId(Context ctx) {
        String idParam = ctx.pathParam("id");
        try {
            return new GiftCardId(UUID.fromString(idParam));
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result("Invalid gift card id format");
            return null;
        }
    }

    private RequestTopUpRequest parseRequestTopUpRequest(Context ctx) {
        try {
            return ctx.bodyAsClass(RequestTopUpRequest.class);
        } catch (Exception ex) {
            ctx.status(400).result("request body is required");
            return null;
        }
    }

    private GiftCardResponse loadGiftCardResponse(Context ctx, GiftCardId cardId) {
        var cardOpt = giftCardQueryService.findById(cardId);
        if (cardOpt.isEmpty()) {
            ctx.status(500).result("gift card not found after update");
            return null;
        }
        return toResponse(cardOpt.get());
    }

    static GiftCardResponse toResponse(GiftCardDetails card) {
        return new GiftCardResponse(
            card.id(),
            card.balance().value()
        );
    }
}
